lib/cmsis_rtos_v1/cmsis_thread.c

```diff
@@ -0,0 +1,246 @@
+/*
+ * Copyright (c) 2018 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <kernel_structs.h>
+#include <atomic.h>
+#include <init.h>
+#include <device.h>
+#include "cmsis_thread.h"
+
+struct work_item {
+	struct cmsis_thread *thrd;
+	struct k_delayed_work work;
+};
+static struct work_item worker;
+
+static void work_handler(struct k_work *work)
+{
+	struct work_item *ti = CONTAINER_OF(work, struct work_item, work);
+
+	/*Free the mempool here */
+	k_mem_pool_free(&ti->thrd->block);
+}
+
+static int init_cmsis_rtos_v1_post_kernel(struct device *unused)
+{
+	ARG_UNUSED(unused);
+
+	k_work_init(&worker.work.work, work_handler);
+	return 0;
+}
+
+SYS_INIT(init_cmsis_rtos_v1_post_kernel,
+		POST_KERNEL, CONFIG_KERNEL_INIT_PRIORITY_DEFAULT);
+
+static s32_t zephyr_to_cmsis_priority(u32_t z_prio)
+{
+	return(osPriorityRealtime - z_prio);
+}
+
+static u32_t cmsis_to_zephyr_priority(s32_t c_prio)
+{
+	return(osPriorityRealtime - c_prio);
+}
+
+static void zephyr_thread_wrapper(void *arg1, void *arg2, void *arg3)
+{
+	osThreadId thread_id;
+	struct cmsis_thread *cm_thread;
+	void * (*fun_ptr)(void *) = arg3;
+
+	fun_ptr(arg1);
+
+	thread_id = osThreadGetId();
+
+	/* The below operation gets a pointer to the parent structure
+	 * of the thread which also has k_mem_block as one of its elements.
+	 * This needs to be used to free the mem pool.
+	 */
+	cm_thread = CONTAINER_OF(thread_id, struct cmsis_thread, thread);
+
+	worker.thrd = cm_thread;
+	k_work_submit(&worker.work.work);
+}
+
+/* Memory pool for cmsis thread space */
+K_MEM_POOL_DEFINE(cmsis_thread_pool, sizeof(struct cmsis_thread),
+		sizeof(struct cmsis_thread), CONFIG_MAX_CMSIS_THREAD_COUNT, 4);
+
+/**
+ * @brief Create a thread.
+ *
+ * Parameters
+ * [in]	thread_def - thread definition referenced with osThread.
+ * [in]	arg - pointer that is passed to the thread function as start argument.
+ *
+ * Returns
+ * thread ID for reference by other functions or NULL in case of error.
+ */
+osThreadId osThreadCreate(const osThreadDef_t *thread_def, void *arg)
+{
+	u32_t prio;
+	struct k_mem_block block;
+	struct cmsis_thread *cm_thread;
+	k_tid_t tid;
+
+	__ASSERT(thread_def->stacksize > 0 &&
+		 thread_def->stacksize <= CONFIG_MAX_STACK_SIZE_THREAD,
+		 "invalid stack size\n");
+
+	if (_is_in_isr()) {
+		return NULL;
+	}
+
+	if (!thread_def->instances) {
+		return NULL;
+	}
+
+	if (*(thread_def->used_inst_ptr) >= thread_def->instances) {
+		return NULL;
+	}
```

Comment:
Are those known runtime failure cases that need to be handled? This looks like you're just trying to validate the user-supplied parameters.  Zephyr style is to do that with an ASSERT and leave production runtime builds free of cruft.

Reply:
thanks for pointing out. Will use ASSERT.